### ISSUE TITLE
Restricted groups for publishers

### ISSUE THREAD

**Issue by seanh** [hypothesis/product-backlog#424]
As a publisher I want to control the membership of a group that anyone can read, but that only members can write to, so that I can limit annotations to my staff, authors, invited experts, etc and not need to moderate the annotations.

**Note**: this is an **Epic** - a big user story that's too big to go into a sprint, it'll need to be fleshed out and split up into smaller stories before we can start working on it.

Problems:

* **Creating new annotations in the client.**

  Normally when you select some text and click Annotate, it creates an annotation 
  in whatever group is currently selected:

  ![Creating an annotation](https://cloud.githubusercontent.com/assets/22498/25382090/201be596-29ae-11e7-903b-53427514428f.gif)
  
  Restricted groups will be the first time that it's possible to be logged in
  and to have a group selected in the client but not be allowed to create
  annotations in that group. (None of the public group, private groups, eLife's
  third-party group, or open groups create this situation.)
  
  When a user selects some text and clicks annotate and the sidebar opens up, as
  in the GIF above, what happens if the selected group in the sidebar is a 
  restricted group that the user isn't allowed to annotate in? We can't create a
  new annotation in that group, as we would normally.
  
  Previous discussion of this issue and suggestions here: <https://github.com/hypothesis/product-backlog/issues/267#issuecomment-296999302>

* **Activity pages**. Currently if you can get to a group's page (e.g. by
  following a link from the `/search` page or the client to the group's page) then 
  you can join the group. When a non-member visits a group's page, it shows them 
  a **Join** button.
  
  For example here I paste in the URL of the Hypothesis Reading group, which I'm
  not a member of. I'm shown a **Join** button. When I join the group
  **the same URL reloads again** but now that I'm a member it shows me the group's
  page. **The group's page and the secret invite URL are the same URL**.
    
  If you can get to the group's page then you can join the group.
  
  With restricted groups, we want a new type of group that users who aren't members
  can **see** the group's page but being able to see the group doesn't mean you 
  can join or write annotations in the group.
  
  We will probably have to rewrite our invite / join mechanism for groups, so that
  having the URL of the group's page doesn't mean you can join the group.

**Comment by ajpeddakotla** [comment]
After discussing this w/ various team members, here's what I think a first iteration at restricted groups for publishers looks like: 

1. Group creation: Groups are created by Hypothesis developers or Hypothesis team members using the admin panel (to start)

2. Group members: Are added by Hypothesis developers or Hypothesis team members (to start). Eventually, we want to give group creators the ability to add users.

3. Joining a group: Users won't have a visible way to join groups. The only way for users to be added to a group is for H team members to add them.
-- The share link for a restricted group should be altered, so that users can still share the existence of these groups, but not be able to join them when they view the group. 

4. Visual distinction: Create a way to identify the group as "read only" @dawariley 

5. Activity pages: Restricted groups will show up in activity pages.

**Comment by seanh** [comment]
I wonder whether we can have open group links like `hypothes.is/groups/xyz123/my-group` that are safe for users to share without giving out membership to the group, while at the same time still having our private groups have the exact same link `hypothes.is/groups/xyz123/my-group` shown in the same place in our UI, but if you share _that_ link then it irrevocably enables everyone to join and write in your private group? (Effectively you just turned your private group into an open group forever.)

That seems confusing to users and likely to lead to mistakes.

So I wonder if we'll need to split out a separate card, to do before restricted groups, that is to improve how group invite and joining works?

@dawariley ^

**Comment by seanh** [comment]
@dawariley I think UX input needed on the UX problem with creating new annotations and private groups, noted in [this comment](https://github.com/hypothesis/product-backlog/issues/424#issue-289242739) above

**Comment by dawachan** [comment]
Thanks @ajpeddakotla @seanh will be picking this up today! I proposed/presented a user flow for this in the original groups demo mocks/prototype I created awhile back. I will dig these up, and review in the light of this current card and attach/modify as necessary. Cheers!

**Comment by dawachan** [comment]
**Scenario:**
When a group is selected where the group type is **restricted** there are several problems we need to solve pertaining to users inability to create/write annotations to: 

**1. Disabling the adder functionality:**
Adder features 'Annotate' and 'Highlight' - since a highlight is technically an 'annotation without text' if the user does not have write permissions on a group we should disable the Adder from appearing.

onClick no adder should be displayed. My initial thought was to persist the 'highlight' functionality but as outlined above adding a highlight to the page/document would require write permissions within a group, which users don't have. 

onClick an existing highlight (added by one of the restricted group members) sidebar opens as it does now. 

We might want to have selecting text within a highlighted area or selecting non highlighted text anywhere within a document that contains annotations open the side bar by default. The rationale behind persisting the 'on text selection' open sidebar action, is because users are familiar with this interaction pattern at present and to remove it could lead to potential confusion. 

**2. Read-only messaging displays beneath the group**

The proposal is to re-use the existing in app messaging banner at the top which is shown when users need to be logged into annotate. Text reads: "This group is read-only."

![restricted read-only group users can view but can t add annotations to this group](https://user-images.githubusercontent.com/5212631/36207027-66b7b784-118c-11e8-8a90-47d072a42082.png)

The message should persist when a restricted group is selected.

**Comment by dawachan** [comment]
**3. Hide the add note button** 

The add note button is displayed in the bucket bar (classic theme) and in-line (eLife clean theme). Since a note is an unanchored annotation, and the user is in a restricted group we should hide the ability to add notes when the user does not have write permissions.

**Comment by dawachan** [comment]
**Activity Pages for Restricted groups:** related to activity pages for Open groups card: #422 @hmstepanek is working on. This will need to be moved to a separate card but adding here for completeness.

![activity page - restricted group](https://user-images.githubusercontent.com/5212631/36210599-bcab1924-1196-11e8-8706-7aa27740941a.png)

The only difference between this screen mock and the Open groups is that instead of 'moderators' or 'members' (TBD) is that we have 'authors'.

Note: as with the Open group - the 'Edit Group' link only appears if you are the admin/creator of the group and you are looking at a group you have edit permissions against, otherwise this link will be hidden.

**Comment by dwhly** [comment]
> If the user does not have write permissions on a group we should disable the Adder from appearing.

> We might want to have selecting text within a highlighted area or selecting non highlighted text anywhere within a document that contains annotations open the side bar by default.

Hmm. Here's the problem as I see it.

1. Clearly if a user comes to a page, with the sidebar closed (but sees that H is loaded on the page) and tries to select and nothing happens, that's bad.  Something obviously needs to happen.

2. One of the most frequent behaviors that we observed early on, was that a surprising number of people read documents by highlighting w/ their mouse.  Also, people select in order to Copy/Paste.  I don't think that changing the behavior of the adder in a read-only situation to open the sidebar makes sense, and might actually be quite annoying to some.  

I'd counterpropose that we have an adder which explains that they're in X group, which is read-only.  "John's group is read-only.  Click here to view annotations or change your group".   Clicking would then open the sidebar.  Eventually we might _also_ offer them to "create a personal note or highlight" instead which would shift them into the right scope for that, and perhaps open the editor, as @heatherstaines has suggested.

@seanh Has pointed out that the existing adder flow when you're logged out is equally bad.  Namely that you're shown options in the adder, but they are essentially lies that take you to a sidebar that then demands you login.

This might be an opportunity for us to have an adder which is more properly responsive to context, and can shortcut you to what you need to do (login for instance) straightaway.

**Comment by seanh** [comment]
In case it's useful context: I think there's actually several closely related issues with creating annotations in Hypothesis right now. I don't think any of these are necessarily problems with **the adder** per se, rather they're problems with the whole UX flow of creating annotations, for example I think the **annotation editor** in the sidebar is implicated in many of these. Here's just some off the top of my head but I think there are more. I've listed them in a separate issue https://github.com/hypothesis/product-backlog/issues/476

**Comment by dawachan** [comment]
If we are going to support annotation behaviour even if only 'Post to me' functionality within the sidebar, then I'm not sure we need to change the current adder behaviour. I see your point @dwhly about introducing 'an adder which explains that they're in X group, which is read-only.' But this could equally be handled by the sidebar messaging and not in the adder itself since the basic functionality of the adder is intact and it doesn't necessarily make sense to overload the adder with read-only messaging. If no annotation behaviour was to be supported then yes, I believe the solution you propose is a viable one.

**Comment by dawachan** [comment]
@dwhly I like the suggestion to make an "adder which is more properly responsive to context, and can shortcut you to what you need to do (login for instance) straightaway" or view read-only annotations (in the logged out context).

**Comment by seanh** [comment]
I think we should avoid doing things where the UI doesn't work so we display a message telling the user why it doesn't work, e.g. an adder or sidebar messaging that explains they're in X group which is read-only. This doesn't best enable the user to do what they're trying to do (create an annotation). So I think the adder (or more generally create annotation flow) that is responsive to context and can shortcut you to where you need to go sounds like a better direction.

**Comment by dawariley** [comment]
I think we're all agreed on this @seanh.

**Comment by dawachan** [comment]
Hypothetical question: if there was no possibility to annotate (i.e. post to me) was disabled in a read-only groups context a) would the adder still exist and b) where would it 'short cut' users to?

**Comment by dawachan** [comment]
In such a scenario, would it be legitimate to include a CTA which opens the sidebar in read-only mode and would it be helpful to message users either before they click the link or after.. that they are unable to annotate and why?

**Comment by dawachan** [comment]
I'd rather not design for a hypothetical scenario, until we decide which use case(s) with respect to annotation behaviour within a restricted/read-only group will be supported. Thanks!

**Comment by dawachan** [comment]
@dwhly and I sync'd about annotation behaviour for Restricted groups functionality. Here's what was figured out:

- Users should still be able to 'Post to only me' within a Restricted group.
- 'Only me' annotations should live outside of the group context and be readily accessible through the scope control e.g introducing 'Only me' as an option to filter sidebar annotations by annotations made only by me. Based on frequency of current annotation behaviour it is proposed that 'Only me' features near the top of the scope control below 'Public'.
- The post to button submit button within a restricted/read-only group allows users to post to 'only me'. 
- Adder annotate and highlight behaviours remain in-tact in the use case of Restricted groups.
- I will explore what, if any messaging is displayed to users including in the 'adder' TBD and the sidebar. Suggestions include: incorporating a watermark in the text field to make the scope of the post to explicit.

I will edit/remove the earlier comment and provide updated designs for this issue. Additional issues may need to be created to reflect moving 'only me' annotations out of the group level to a standalone 'only me' category. Referenced here already: #200

**Comment by dawachan** [comment]
Note: the revised implementation of 'Only Me' assumes that private annotations will be stored on the  same namespace for the account the user is currently authenticated against.

e.g. Hypothesis is implemented for a publisher on a third party namespace, and a user is logged into that publisher's annotation service 'Only me' annotations will be stored against the third party namespace user account. 

The interesting question is if a user is logged in/authenticated against multiple accounts and given that 'Only me' annotations will live outside/beyond the scope of a group - which user account will take precedence? Perhaps we need to introduce the concept of which account is the authoritative one and should this be under the user's control? (Similar to how Google handles account authorities).

**Comment by dawachan** [comment]
I've spoken at length with @seanh this morning and it might make sense for v1 to switch the scope of the groups menu to 'Only me' when the user clicks 'Annotate'. I think this is worth exploring, but I need a little more time to think this through as we might need to introduce a short cut to view the Restricted group annotations associated with the current page/content within the Adder or a shortcut to view this within the sidebar. Caveat: we might not be able to solve all these problems in the v1 implementation of restricted groups.

**Comment by dawachan** [comment]
The alternative is to persist the user on the group they currently have selected or the default group scoped for the page (e.g Restricted Publisher group) and have the 'Post to Me' annotation editor create an annotation which doesn't persist on the screen after the submit button is pressed. A user would then have to navigate to 'Only me' to view the annotation they just created. I need some time to explore these options.

**Comment by hmstepanek** [comment]
In regards to 

> 5. Activity pages: Restricted groups will show up in activity pages.

I don't see any reasons why this wouldn't already work. (Maybe I'm missing something.) The bulk of the activity page work will be the join link mechanism. 

> 3. Joining a group: Users won't have a visible way to join groups. The only way for users to be added to a group is for the group creator to add them.
-- The share link for a restricted group should be altered, so that users can still share the existence of these groups, but not be able to join them when they view the group.

I'm thinking (and again I could be missing something), that if the group is readable like the open group then it will not redirect the user to the join group page. Regardless of current behavior, we really just need to not redirect them and take them straight to the activity page minus any join links so shouldn't be to tricky.

**Comment by ajpeddakotla** [comment]
Broke up this ticket into several cards detailed here: 

- [ ] Admin Interface for creating open and restricted scoped groups: #421 
- [ ] Add group members to resticted groups: #494
- [ ] Activity Pages for Restricted Groups: #495
- [ ] Add restricted groups for publishers to Hypothesis groups model #496


NOTE: 
The additional requirements that were identified as part of this card, but will not be tackled in this first sprint for restricted groups - the UX changes to current annotation flow - have been detailed in this card: #489

**Comment by hmstepanek** [comment]
@ajpeddakotla Can we close this?
